# Spark SQL with pySpark

Find out how many orders, how many products and how many sellers are in the data.

In [ ]:
from pyspark.sql import SparkSession
from pyspark.sql.functions import *

In [ ]:
spark = SparkSession.builder \
    .master("local") \
    .config("spark.sql.autoBroadcastJoinThreshold", -1) \
    .config("spark.executor.memory", "500mb") \
    .appName("Exercise1") \
    .getOrCreate()

In [ ]:
# Read the source tables

products_table = spark.read.parquet("./data/products_parquet")
sales_table = spark.read.parquet("./data/sales_parquet")
sellers_table = spark.read.parquet("./data/sellers_parquet")

# Databricks 

# products_table = spark.read.parquet("s3://codeop-react/ds/datasets/spark/practice_1/products_parquet/unified_products.parquet")
# sales_table = spark.read.parquet("s3://codeop-react/ds/datasets/spark/practice_1/sales_parquet/unified_sales.parquet")
# sellers_table = spark.read.parquet("s3://codeop-react/ds/datasets/spark/practice_1/sellers_parquet/unified_sellers.parquet")

In [ ]:
sales_table.columns


Out[27]: ['order_id',
 'product_id',
 'seller_id',
 'date',
 'num_pieces_sold',
 'bill_raw_text']

In [ ]:
sales_table.printSchema()

root
 |-- order_id: string (nullable = true)
 |-- product_id: string (nullable = true)
 |-- seller_id: string (nullable = true)
 |-- date: string (nullable = true)
 |-- num_pieces_sold: string (nullable = true)
 |-- bill_raw_text: string (nullable = true)



1. Print the number of orders

Number of Orders: 100195


Number of Orders: 100195

2. Print the number of sellers

Number of sellers: 10


Number of sellers: 10

3. Print the number of products

Number of products: 9570548


Number of products: 9570548

4. Number of products sold at least once

+-----------------+
|count(product_id)|
+-----------------+
|            99513|
+-----------------+



result: 99513

5. Product present in more orders

+----------+-----+
|product_id|count|
+----------+-----+
|  31136332|    3|
+----------+-----+



| product_id | count    |
|------------|----------|
| 31136332   |    3     |

6. How many distinct products have been sold in each date

|       date | distinct_products_sold |
|-----------:|------------------------|
| 2020-07-09 | 10137                  |
| 2020-07-01 | 10136                  |
| 2020-07-04 | 10032                  |
| 2020-07-06 | 10030                  |
| 2020-07-03 | 10022                  |
| 2020-07-08 | 10021                  |
| 2020-07-02 | 10003                  |
| 2020-07-07 | 9980                   |
| 2020-07-05 | 9967                   |
| 2020-07-10 | 9801                   |

7. What is the average revenue of the orders?

+-----------------------+
|average_revenue x order|
+-----------------------+
|     3827.7561790766363|
+-----------------------+



Result: 3827.7561790766363